src/security-council-mgmt/governors/modules/SecurityCouncilMemberElectionGovernorCountingUpgradeable.sol

```diff
@@ -0,0 +1,315 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts-upgradeable/governance/GovernorUpgradeable.sol";
+
+/// @title AccountRankerUpgradeable
+/// @notice Keeps track of the top K nominees for a given proposalId and their weights
+abstract contract AccountRankerUpgradeable is Initializable {
+    /// @dev max number of nominees to track (6)
+    uint256 private _maxNominees;
+
+    /// @dev proposalId => list of top nominees in descending order by weight
+    mapping(uint256 => address[]) private _nominees;
```

Comment:
this might be a little out there, but what if we could avoid rolling our own algorithm by using solady's quicksort? something like this maybe? i have no idea how much gas this would cost with 500 nominees, but i would guess it's within the max gas.

```solidity
import "lib/solady/src/utils/LibSort.sol";
contract Foo {
    address[] nomineesWithVotes; // list of nominees who've received votes
    mapping(address => uint96) weightReceived; // weight received by nominee (max votes is ~80 bil * 1e18)
    
    function getTopNominees() public view returns (address[] memory) {
        uint256[] memory nomineesToSort = new uint256[](nomineesWithVotes.length);
        
        for (uint256 i = 0; i < nomineesWithVotes.length; i++) {
            address nominee = nomineesWithVotes[i];
            nomineesToSort[i] = (uint256(weightReceived[nominee]) << 160) & uint160(nominee);
        }

        LibSort.sort(nomineesToSort);

        address[] memory topNominees = new address[](6);

        for (uint256 i = 0; i < 6; i++) {
            topNominees[i] = address(uint160(nomineesToSort[i]));
        }

        return topNominees;
    }
}
```

Reply:
i think the last for loop would need to be reversed but that's the idea at least

Reply:
Yeh, this is pretty neat. But I'm a bit wary of that max, not sure we're constricted to 80b elsewhere are we? We might be by some of the bravo compatibility. To make the max bigger you could use the index of the nominee in the nomineesWithVotes (which wouldnt need many bits), rather than the address of the nominee.

We could also be more efficient by only keeping track of the six highest nominees. Then foreach nominee in the large list we check if it's higher than the current lowest, if so add it to the highest 6 and remove the last item in the 6.

Reply:
something like this perhaps?

```solidity
import "lib/solady/src/utils/LibSort.sol";
contract Foo {
    address[] nomineesWithVotes; // list of nominees who've received votes
    mapping(address => uint240) weightReceived; // weight received by nominee

    function two() public view returns (address[] memory) {
        uint256[] memory topNominees = new uint256[](6);
        uint256 topNomineesLength = 0;

        for (uint16 i = 0; i < nomineesWithVotes.length; i++) {
            uint256 packed = (uint256(weightReceived[nomineesWithVotes[i]]) << 16) & i;

            if (topNomineesLength < 5) {
                topNominees[topNomineesLength] = packed;
                topNomineesLength++;
            } 
            else if (topNomineesLength == 5) {
                topNominees[topNomineesLength] = packed;
                topNomineesLength++;
                LibSort.insertionSort(topNominees);
            }
            else if (topNominees[0] < packed) {
                topNominees[0] = packed;
                LibSort.insertionSort(topNominees);
            }
        }

        address[] memory topNomineesAddresses = new address[](6);

        for (uint256 i = 0; i < 6; i++) {
            topNomineesAddresses[i] = nomineesWithVotes[uint16(topNominees[i])];
        }
    }
}
```

Reply:
solady insertionSort is also way simpler than the quicksort implementation. more efficient for small arrays too!

Reply:
something like this maybe?

```solidity
import "lib/solady/src/utils/LibSort.sol";
contract Foo {
    address[] nomineesWithVotes; // list of nominees who've received votes
    mapping(address => uint240) weightReceived; // weight received by nominee

    function set(address nominee, uint240 weight) public {
        weightReceived[nominee] = weight;
        nomineesWithVotes.push(nominee);
    }

    function two() public view returns (address[] memory) {
        uint256[] memory topNominees = new uint256[](6);
        uint256 topNomineesLength = 0;

        for (uint16 i = 0; i < nomineesWithVotes.length; i++) {
            address nominee = nomineesWithVotes[i];
            uint256 weight = weightReceived[nominee];
            uint256 packed = (weight << 16) | i;

            if (topNomineesLength < 5) {
                topNominees[topNomineesLength] = packed;
                topNomineesLength++;
            } 
            else if (topNomineesLength == 5) {
                topNominees[topNomineesLength] = packed;
                topNomineesLength++;
                LibSort.insertionSort(topNominees);
            }
            else if (topNominees[0] < packed) {
                topNominees[0] = packed;
                LibSort.insertionSort(topNominees);
            }
        }

        address[] memory topNomineesAddresses = new address[](6);

        for (uint256 i = 0; i < 6; i++) {
            topNomineesAddresses[i] = nomineesWithVotes[uint16(topNominees[i])];
        }

        return topNomineesAddresses;
    }
}

contract FooTest is Test {
    Foo f;

    function setUp() public {
        f = new Foo();

        // make the worst case array, 500 in ascending order
        for (uint256 i = 0; i < 500; i++) {
            f.set(address(uint160(i + 1)), uint240(i + 1));
        }
    }

    function testFoo() public {
        // see how much gas it takes to sort
        uint256 g = gasleft();
        address[] memory result = f.two();
        g = g - gasleft();

        console.log(g); // about 3 mil gas

        assertEq(result[5], address(500));
        assertEq(result[4], address(499));
        assertEq(result[3], address(498));
        assertEq(result[2], address(497));
        assertEq(result[1], address(496));
        assertEq(result[0], address(495));
    }
}
```

Reply:
Yeh, I had imagined doing it more custom and just iterating an array of structs and moving them around. But I guess this works too. I'd have to look closer at your bit sizes in this to be sure it's good, like is this correct?
`uint256 packed = (weight << 16) | i;`.

Reply:
i think so? i wouldn't be surprised if it's wrong though